src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,402 @@
+        private static readonly MemberInfo CompletedTaskMemberInfo = GetMemberInfo<Func<Task>>(() => Task.CompletedTask);
+
+        private static readonly ParameterExpression TargetArg = Expression.Parameter(typeof(object), "target");
+        private static readonly ParameterExpression HttpContextParameter = Expression.Parameter(typeof(HttpContext), "httpContext");
+        private static readonly ParameterExpression DeserializedBodyArg = Expression.Parameter(typeof(object), "bodyValue");
+
+        private static readonly MemberExpression RequestServicesExpr = Expression.Property(HttpContextParameter, nameof(HttpContext.RequestServices));
+        private static readonly MemberExpression HttpRequestExpr = Expression.Property(HttpContextParameter, nameof(HttpContext.Request));
+        private static readonly MemberExpression HttpResponseExpr = Expression.Property(HttpContextParameter, nameof(HttpContext.Response));
+
+        public static RequestDelegate BuildRequestDelegate(Delegate action)
+        {
+            // Non void return type
+
+            // Task Invoke(HttpContext httpContext)
+            // {
+            //     // Action parameters are bound from the request, services, etc... based on attribute and type information.
+            //     return ExecuteTask(action(...), httpContext);
+            // }
+
+            // void return type
+
+            // Task Invoke(HttpContext httpContext)
+            // {
+            //     action(...);
+            //     return default;
+            // }
+
+            var method = action.Method;
+
+            var consumeBodyDirectly = false;
+            var consumeBodyAsForm = false;
+            Type? bodyType = null;
+            var allowEmptyBody = false;
+
+            // This argument represents the deserialized body returned from IHttpRequestReader
+            // when the method has a FromBody attribute declared
+
+            var args = new List<Expression>();
+
+            foreach (var parameter in method.GetParameters())
+            {
+                Expression paramterExpression = Expression.Default(parameter.ParameterType);
+
+                if (parameter.GetCustomAttributes().OfType<IFromRouteMetadata>().FirstOrDefault() is { } routeAttribute)
+                {
+                    var routeValuesProperty = Expression.Property(HttpRequestExpr, nameof(HttpRequest.RouteValues));
+                    paramterExpression = BindParamenter(routeValuesProperty, parameter, routeAttribute.Name);
+                }
+                else if (parameter.GetCustomAttributes().OfType<IFromQueryMetadata>().FirstOrDefault() is { } queryAttribute)
+                {
+                    var queryProperty = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Query));
+                    paramterExpression = BindParamenter(queryProperty, parameter, queryAttribute.Name);
```

Comment:
Is there a plan to support optional parameter?

Reply:
The parameter binding is incomplete. Being an MVP, there are still quite a few rough edges. I've added a comment to https://github.com/dotnet/aspnetcore/issues/30248 to track this though.